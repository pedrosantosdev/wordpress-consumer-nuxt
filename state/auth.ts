import { defineStore, acceptHMRUpdate } from 'pinia'
import { AuthModel } from '@/types/User'
import { isNotEmpty } from '@/helpers/string'
import { useBaseFetch } from '@/composables/useBaseFetch'
import { navigateTo } from '#app'

type ResponseAuth = {
	accessToken: string
	refreshToken: string
	expires: string
}

export type AuthState = {
	lastPage: string
	expiresAt: string | null
	error: {
		message: string | null
		code: string | null
	}
	onRequest: boolean
} & AuthModel

export const useAuthStore = defineStore({
	id: 'auth',
	state: (): AuthState => {
		return {
			lastPage: '/',
			token: null,
			refreshToken: null,
			expiresAt: null,
			user: null,
			onRequest: false,
			error: {
				message: null,
				code: null,
			},
		}
	},
	actions: {
		setError(message: string | null = null, code: string | null = null): void {
			this.$patch((state: AuthState) => {
				state.error = {
					message,
					code,
				}
			})
		},
		async setCredential(response: ResponseAuth) {
			if (
				[response.accessToken, response.refreshToken, response.expires].every((value) =>
					isNotEmpty(value)
				)
			) {
				await this.$patch((state: AuthState) => {
					state.token = response.accessToken ?? ''
					state.refreshToken = response.refreshToken ?? ''
					state.expiresAt = response.expires ?? ''
				})
			}
		},
		async login(payload: FormData) {
			if (
				!isNotEmpty(payload.get('username') as string) ||
				!isNotEmpty(payload.get('password') as string)
			) {
				this.setError('Missing Fields', 'empty_fields')
				return
			}
			this.setError()
			const response = await useBaseFetch<ResponseAuth>(`login`, {
				method: 'POST',
				body: payload,
				headers: {
					Accept: 'application/json',
				},
			})
			if (!response || response.error.value) {
				this.setError('Invalid Credentials', 'invalid_credentials')
				return
			}
			this.setCredential(response.data.value as ResponseAuth)
		},
		async refreshToken() {
			if (!this.$state.refreshToken || !this.$state.token) {
				return false
			}
			this.$state.onRequest = true
			const response = await useBaseFetch<ResponseAuth>(`refresh`, {
				method: 'POST',
				body: {
					access_token: this.$state.token,
					refresh_token: this.$state.refreshToken,
				},
				headers: {
					Accept: 'application/json',
				},
			})
			this.$state.onRequest = false
			if (!response || response.error.value) {
				this.setError('Invalid Credentials', 'invalid_credentials')
				return false
			}
			this.setCredential(response.data.value as ResponseAuth)
			return true
		},
		async logout(redirect = true) {
			await this.$reset()
			if (redirect) {
				return navigateTo('login')
			}
		},
	},
	getters: {
		hasError: (state: AuthState) => isNotEmpty(state.error.message) || isNotEmpty(state.error.code),
		isAuth: (state: AuthState) => isNotEmpty(state.token),
		isExpired: (state: AuthState) =>
			state.expiresAt && state.refreshToken
				? new Date(Date.parse(state.expiresAt)) < new Date()
				: true,
	},
	persist: {
		paths: ['token', 'expiresAt', 'refreshToken'],
	},
})

if (import.meta.hot) {
	import.meta.hot.accept(acceptHMRUpdate(useAuthStore, import.meta.hot))
}
