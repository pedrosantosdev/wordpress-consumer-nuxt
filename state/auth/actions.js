import { useBaseFetch } from '@/composables/baseFetch'

const actions = {
  async login(payload) {
    if (this.isLoading) {
      return
    } else if (payload.username === '' || payload.password === '') {
      this.error = {
        message: 'Missing Fields',
        code: 'empty_fields',
      }
      return
    }
    this.isLoading = true
    this.error = {
      message: '',
      code: '',
    }
    await useBaseFetch(`login`, {
      method: 'POST',
      body: payload,
      headers: {
        Accept: 'application/json',
      },
    })
      .then((response) => {
        this.user = { name: payload.username }
        this.token = response.token
      })
      .catch(
        (response) =>
          (this.error = {
            message: response.data,
            code: response.code,
          })
      )
      .finally(() => (this.isLoading = false))
  },
}

export default actions
