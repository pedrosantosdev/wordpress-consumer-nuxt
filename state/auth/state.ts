import { AuthModel } from '@/types/User'

export interface AuthState extends AuthModel {
  lastPage: string
  isLoading: false
  error: {
    message: string
    code: string
  }
}

const state = (): AuthState => ({
  lastPage: 'index',
  isAuth: false,
  token: null,
  user: null,
  isLoading: false,
  error: {
    message: '',
    code: '',
  },
})

export default state
