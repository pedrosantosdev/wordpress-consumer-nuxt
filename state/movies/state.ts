import { MovieModel } from '@/types/Movie'

export interface StateModel {
  list?: MovieModel[]
  isLoading: boolean
  hasError: boolean
  query: string
  queryResultList?: MovieModel[]
}

const state = (): StateModel => ({
  list: [],
  isLoading: false,
  hasError: false,
  query: '',
  queryResultList: [],
})

export default state