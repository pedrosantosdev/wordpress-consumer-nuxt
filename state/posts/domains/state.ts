import { PostDomain } from '@/types/Post'

export interface StateModel {
	list?: PostDomain[]
	isLoading: boolean
	hasError: boolean
}

const state = (): StateModel => ({
	list: [],
	isLoading: false,
	hasError: false,
})

export default state
