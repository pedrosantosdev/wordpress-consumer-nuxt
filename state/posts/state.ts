import { Post } from '@/types/Post'
import { Page } from '@/types/Page'

export interface StateModel {
	list?: Post[]
	searchList: Page<Post> | null
	isLoadingSearch: boolean
	currentPost: Post | null
	isLoading: boolean
	hasError: boolean
}

const state = (): StateModel => ({
	list: [],
	searchList: null,
	isLoading: false,
	hasError: false,
	currentPost: null,
	isLoadingSearch: false,
})

export default state
