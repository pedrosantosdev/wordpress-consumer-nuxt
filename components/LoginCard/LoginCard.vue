<template>
	<BaseCard>
		<div class="self-center text-xl font-light text-gray-600 sm:text-2xl dark:text-white">
			Login To Your Account
		</div>
		<div class="mt-2">
			<form autoComplete="off" :class="{ error: hasError }" @submit.prevent="onSubmit">
				<div class="flex flex-col mb-2">
					<div class="flex relative flex-wrap input-group">
						<label for="username" class="prefix-input">
							<NuxtIcon name="mail" />
						</label>
						<input
							id="username"
							v-model.trim="form.username.value"
							type="text"
							name="username"
							placeholder="Your email"
						/>
						<NuxtIcon name="times" class="error--icon" />
					</div>
				</div>
				<div class="flex flex-col mb-4">
					<div class="flex relative flex-wrap input-group">
						<label for="password" class="prefix-input">
							<NuxtIcon name="unlock" />
						</label>
						<input
							id="password"
							v-model.trim="form.password.value"
							type="password"
							name="password"
							placeholder="Your password"
						/>
						<NuxtIcon name="times" class="error--icon" />
					</div>
					<p class="text-sm text-red-500 -bottom-6 w-full error--message">Invalid Credentials</p>
				</div>
				<div class="flex w-full">
					<BaseButton :disabled="isLoading" :show-icon="isLoading" @click="onSubmit"
						>Login</BaseButton
					>
				</div>
			</form>
		</div>
	</BaseCard>
</template>

<script setup lang="ts">
import { useAuthStore } from '@/state/auth'
import { storeToRefs } from 'pinia'
import { ref, onBeforeMount } from 'vue'
import { navigateTo } from 'nuxt/app'

const authStore = useAuthStore()
const { hasError, isAuth, lastPage, isExpired } = storeToRefs(authStore)
const isLoading = ref(false)
const form = {
	username: ref(''),
	password: ref(''),
}

onBeforeMount(() => {
	if (isAuth.value && isExpired.value) {
		authStore.logout(false)
	} else if (isAuth.value) {
		goForward()
	}
})
authStore.$subscribe(() => {
	if (isAuth.value) {
		goForward()
	}
})
function goForward(): void {
	navigateTo(lastPage.value ?? '/')
}
async function onSubmit(): Promise<void> {
	const data = new FormData()
	data.append('username', form.username.value)
	data.append('password', form.password.value)
	isLoading.value = true
	await authStore.login(data)
	isLoading.value = false
}
</script>

<style lang="scss" scoped>
.error {
	&--message,
	&--icon {
		@apply opacity-0 block transition-all;
	}
	&--icon {
		top: 12px;
		right: 10px;
		@apply text-red-500 absolute;
	}
}
.input-group.input-group {
	&:focus-within {
		input,
		.prefix-input {
			@apply border-purple-600;
		}
	}
	input,
	.prefix-input {
		@apply bg-white border-gray-300 shadow-sm border outline-none border-2;
	}
	.prefix-input {
		@apply rounded-l-md inline-flex items-center px-3 border-r-0 text-gray-500 text-sm border-r-gray-300;
	}
	input {
		@apply rounded-r-md flex-1 appearance-none border-l-0 w-full py-2 px-4 text-gray-700 placeholder-gray-400 text-base border-l-gray-300;
	}
}
.error {
	.error--message,
	.error--icon {
		@apply opacity-100;
	}
	.input-group {
		input,
		.prefix-input {
			@apply border-red-500;
		}
	}
}
</style>
