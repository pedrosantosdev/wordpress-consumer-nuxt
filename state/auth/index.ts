import { defineStore, acceptHMRUpdate } from 'pinia'
import state from './state'
import actions from './actions'
import getters from './getters'

export const useAuthStore = defineStore('auth', {
  state,
  actions,
  getters,
  persist: {
    paths: ['token', 'refreshToken', 'expiresAt'],
  },
})

if (import.meta.hot) {
  import.meta.hot.accept(acceptHMRUpdate(useAuthStore, import.meta.hot))
}
