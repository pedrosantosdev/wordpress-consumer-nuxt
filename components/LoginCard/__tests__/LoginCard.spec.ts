import { describe, it, expect, vi, afterEach } from 'vitest'
import { fireEvent, render } from '@testing-library/vue'
import LoginCard from '@/components/LoginCard/LoginCard.vue'
import { createTestingPinia } from '@pinia/testing'
import { useAuthStore } from '@/state/auth'
import type { StateTree } from 'pinia'
import nuxtMock from '@/test/mock/FakeNuxt'

const createPiniaSut = (state: StateTree | undefined = undefined) =>
	createTestingPinia({ stubActions: false, initialState: state })
describe('LoginCard', () => {
	afterEach(() => {
		vi.resetAllMocks()
	})
	let initialState = undefined
	it('should render properly', () => {
		const { html } = render(LoginCard, {
			global: {
				stubs: {
					NuxtIcon: true,
				},
				plugins: [createPiniaSut()],
			},
		})
		expect(html()).toMatchSnapshot()
	})
	it('should call logout when isAuth and Expired', () => {
		initialState = {
			auth: {
				token: 'token',
				refreshToken: 'refreshToken',
				expiresAt: '2020-01-01 10:00:00',
			},
		}
		render(LoginCard, {
			global: {
				stubs: {
					NuxtIcon: true,
				},
				plugins: [createPiniaSut(initialState)],
			},
		})
		const authStore = useAuthStore()
		expect(authStore.logout).toBeCalledTimes(1)
		expect(authStore.isAuth).toBeFalsy()
	})
	it('should redirect when isAuth and not Expired', () => {
		initialState = {
			auth: {
				token: 'token',
				refreshToken: 'refreshToken',
				expiresAt: '2050-01-01T10:00:00',
			},
		}
		render(LoginCard, {
			global: {
				stubs: {
					NuxtIcon: true,
				},
				plugins: [createPiniaSut(initialState)],
			},
		})
		const authStore = useAuthStore()
		expect(authStore.logout).not.toBeCalled()
		expect(authStore.isAuth).toBeTruthy()
		expect(nuxtMock.navigateTo).toBeCalledTimes(1)
	})
	it('should redirect when subscribe isAuth', async () => {
		render(LoginCard, {
			global: {
				stubs: {
					NuxtIcon: true,
				},
				plugins: [createPiniaSut()],
			},
		})
		const authStore = useAuthStore()
		await authStore.$patch((state) => (state.token = 'token'))
		expect(authStore.logout).not.toBeCalled()
		expect(authStore.isAuth).toBeTruthy()
		expect(nuxtMock.navigateTo).toBeCalledTimes(1)
	})
	it('should call login action when click button form', async () => {
		const { getByText } = render(LoginCard, {
			global: {
				stubs: {
					NuxtIcon: true,
				},
				plugins: [createPiniaSut()],
			},
		})
		const authStore = useAuthStore()
		await fireEvent.click(getByText('Login'))
		expect(authStore.login).toBeCalledTimes(1)
	})
})
