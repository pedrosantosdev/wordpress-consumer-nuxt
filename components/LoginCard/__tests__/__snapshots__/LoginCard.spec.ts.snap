// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginCard > should render properly 1`] = `
"<div data-v-47574c68="" class="lg:w-96 w-full sm:h-64 md:h-auto flex flex-col sm:flex-row p-2 sm:p-4 rounded-2xl shadow-lg gap-5 select-none dark:bg-gray-800 bg-white dark:bg-gray-800 dark:text-white">
  <div data-v-47574c68="" class="self-center text-xl font-light text-gray-600 sm:text-2xl dark:text-white">form.title</div>
  <div data-v-47574c68="" class="mt-2">
    <form data-v-47574c68="" autocomplete="off" class="">
      <div data-v-47574c68="" class="flex flex-col mb-2">
        <div data-v-47574c68="" class="flex relative flex-wrap input-group"><label data-v-47574c68="" for="username" class="prefix-input">
            <nuxt-icon-stub data-v-47574c68="" name="mail"></nuxt-icon-stub>
          </label><input data-v-47574c68="" id="username" type="text" name="username" placeholder="form.placeholder.email">
          <nuxt-icon-stub data-v-47574c68="" name="times" class="error--icon"></nuxt-icon-stub>
        </div>
      </div>
      <div data-v-47574c68="" class="flex flex-col mb-4">
        <div data-v-47574c68="" class="flex relative flex-wrap input-group"><label data-v-47574c68="" for="password" class="prefix-input">
            <nuxt-icon-stub data-v-47574c68="" name="unlock"></nuxt-icon-stub>
          </label><input data-v-47574c68="" id="password" type="password" name="password" placeholder="form.placeholder.password">
          <nuxt-icon-stub data-v-47574c68="" name="times" class="error--icon"></nuxt-icon-stub>
        </div>
        <p data-v-47574c68="" class="text-sm text-red-500 -bottom-6 w-full error--message">form.invalid.message</p>
      </div>
      <div data-v-47574c68="" class="flex w-full"><button data-v-47574c68="" type="button" class="py-2 px-4 h-10 bg-purple-600 hover:bg-purple-700 focus:ring-purple-500 focus:ring-offset-purple-200 text-white w-full transition ease-in duration-200 text-center text-base font-semibold shadow-md focus:outline-none focus:ring-2 focus:ring-offset-2 rounded-lg">form.submit.button</button></div>
    </form>
  </div>
</div>"
`;
